src/expression.c

```diff
@@ -6104,8 +6104,9 @@ Expression *TypeidExp::semantic(Scope *sc)
     {
         /* Get the dynamic type, which is .classinfo
          */
-        e = new DotIdExp(ea->loc, ea, Id::classinfo);
-        e = e->semantic(sc);
+        ea = ea->semantic(sc);
+        e = new TypeidExp(ea->loc, ea);
+        e->type = Type::typeinfoclass->type;
```

Comment:
If I remember correctly, `typeid(obj)` and `obj.classinfo` have different behaviors when the expression `obj` is an `interface` reference. Are you sure that it does not change anything?